[Site Isolation] Choose correct web process for scrolling updates from ui process

nmoucht opened the pull request:
#### 733824ab0459c4af6ebd6dbd80fa64d1cf86ddd1
<pre>
[Site Isolation] Choose correct web process for scrolling updates from ui process
<a href="https://bugs.webkit.org/show_bug.cgi?id=268602">https://bugs.webkit.org/show_bug.cgi?id=268602</a>
<a href="https://rdar.apple.com/122162277">rdar://122162277</a>

Reviewed by Alex Christensen.

This is part 3 of splitting up #23242
This chooses the correct web process to send scrolling updates to
based on a map of process ids to root frame ids.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::frameViewForScrollingNode const):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::frameIDForScrollingNodeID):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::sendScrollingTreeNodeDidScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendScrollPositionChangedForNode):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/274932@main">https://commits.webkit.org/274932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebaba8edf0eea1dc68526f314d6ee0369c32b05c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40422 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42800 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42969 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36511 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42729 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22386 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16763 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33531 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16388 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14115 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14181 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35819 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44245 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36649 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39884 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15237 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12480 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38171 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16856 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16905 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5357 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16500 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -485,6 +486,9 @@ LocalFrameView* AsyncScrollingCoordinator::frameViewForScrollingNode(ScrollingNo
     
     RefPtr localMainFrame = dynamicDowncast<LocalFrame>(page()->mainFrame());
+    if (!localMainFrame && is<RemoteFrame>(page()->mainFrame()))
+        localMainFrame = dynamicDowncast<LocalFrame>(page()->mainFrame().tree().firstChild());
```
This is not correct.  It is not the main frame, and there is nothing special about the first child frame.  It looks like we need to iterate the Page's root frames instead.

nmoucht replied:
Updated this to iterate through `page()->rootFrames()`.

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -516,6 +512,26 @@ LocalFrameView* AsyncScrollingCoordinator::frameViewForScrollingNode(ScrollingNo
+    RefPtr localMainFrame = dynamicDowncast<LocalFrame>(page()->mainFrame());
+    if (localMainFrame)
+        return frameViewForScrollingNode(*localMainFrame, scrollingNodeID);
+    if (!localMainFrame && is<RemoteFrame>(page()->mainFrame())) {
```
All this is probably unnecessary.  With site isolation off, and in the main frame process, you can count on Page::rootFrames() to return the main frame.  If the main frame is local and there's also a local grandchild frame that is also a root frame, then this code is actually incorrect.

nmoucht replied:
Right fixed.

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -76,6 +76,7 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
     WEBCORE_EXPORT void setMouseIsOverContentArea(ScrollableArea&, bool) override;
     WEBCORE_EXPORT void setMouseMovedInContentArea(ScrollableArea&) override;
     WEBCORE_EXPORT void setLayerHostingContextIdentifierForFrameHostingNode(ScrollingNodeID, std::optional<LayerHostingContextIdentifier>) override;
+    LocalFrameView* frameViewForScrollingNode(LocalFrame& localMainFrame, ScrollingNodeID) const;
```
This parameter name should be changed to rootFrame.  You might also want to add ASSERT(rootFrame.isRootFrame()) to make that more clear in the code and verify in debug builds.

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -516,6 +513,21 @@ LocalFrameView* AsyncScrollingCoordinator::frameViewForScrollingNode(ScrollingNo
+
+    for (const WeakRef<LocalFrame>& rootFrameRef : page()->rootFrames()) {
+        if (auto* rootFrame = rootFrameRef.ptr()) {
+            if (auto* frameView = frameViewForScrollingNode(*rootFrame, scrollingNodeID))
```
if (!page())
    return nullptr;
for (const auto& rootFrame : page()->rootFrames()) {
    if (RefPtr frameView = frameViewForScrollingNode(rootFrame.get(), scrollingNodeID))

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -516,6 +513,18 @@ LocalFrameView* AsyncScrollingCoordinator::frameViewForScrollingNode(ScrollingNo
     }
 
     return nullptr;
+
```
nit: extra space.

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -500,12 +500,9 @@ LocalFrameView* AsyncScrollingCoordinator::frameViewForScrollingNode(ScrollingNo
     // Walk the frame tree to find the matching LocalFrameView. This is not ideal, but avoids back pointers to LocalFrameViews
     // from ScrollingTreeStateNodes.
-    for (auto* frame = &page()->mainFrame(); frame; frame = frame->tree().traverseNext()) {
+    for (Frame* frame = &rootFrame; frame; frame = frame->tree().traverseNext()) {
```
Could we make this RefPtr<Frame>?

nmoucht replied:
Fixed.

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -1008,6 +1008,17 @@ void ScrollingTree::addScrollingNodeToHostedSubtreeMap(LayerHostingContextIdenti
+{
+    Locker locker { m_treeLock };
+    auto it = m_nodeMap.find(scrollingNodeID);
+    if (it != m_nodeMap.end()) {
```
Less indentation if we early return if it is the end.

nmoucht committed:
[Site Isolation] Choose correct web process for scrolling updates from ui process
https://bugs.webkit.org/show_bug.cgi?id=268602
rdar://122162277

Reviewed by Alex Christensen.

This is part 3 of splitting up #23242
This chooses the correct web process to send scrolling updates to
based on a map of process ids to root frame ids.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::frameViewForScrollingNode const):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::frameIDForScrollingNodeID):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::sendScrollingTreeNodeDidScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendScrollPositionChangedForNode):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/274932@main
